x/merkledb/intermediate_node_db.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package merkledb
+
+import (
+	"sync"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+const (
+	defaultBufferLength       = 256
+	intermediateNodePrefixLen = 1
+)
+
+var intermediateNodePrefix = []byte{2}
+
+type intermediateNodeDB struct {
+	// Holds unused []byte
+	bufferPool *sync.Pool
+
+	// The underlying storage
+	baseDB database.Database
+
+	// If a value is nil, the corresponding key isn't in the trie.
+	// Note that a call to Put may cause a node to be evicted
+	// from the cache, which will call [OnEviction].
+	// A non-nil error returned from Put is considered fatal.
+	nodeCache         onEvictCache[path, *node]
+	evictionBatchSize int
+	metrics           merkleMetrics
+}
+
+func newIntermediateNodeDB(db database.Database, bufferPool *sync.Pool, metrics merkleMetrics, size int, evictionBatchSize int) *intermediateNodeDB {
+	result := &intermediateNodeDB{
+		metrics:           metrics,
+		baseDB:            db,
+		bufferPool:        bufferPool,
+		evictionBatchSize: evictionBatchSize,
+	}
+	result.nodeCache = newOnEvictCache[path](size, result.onEviction)
+	return result
+}
+
+func (db *intermediateNodeDB) prefixedKey(key []byte) []byte {
+	prefixedKey := db.bufferPool.Get().([]byte)
+	keyLen := intermediateNodePrefixLen + len(key)
+	if cap(prefixedKey) >= keyLen {
+		// The [] byte we got from the pool is big enough to hold the prefixed key
+		prefixedKey = prefixedKey[:keyLen]
+	} else {
+		// The []byte from the pool wasn't big enough.
+		// Put it back and allocate a new, bigger one
+		db.bufferPool.Put(prefixedKey)
+		prefixedKey = make([]byte, keyLen)
+	}
+	copy(prefixedKey, intermediateNodePrefix)
+	copy(prefixedKey[intermediateNodePrefixLen:], key)
+	return prefixedKey
+}
```

Comment:
Can we remove some of the duplicate code here? this is almost exactly the same as the value node db function.